terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java

```diff
@@ -0,0 +1,2447 @@
+                        codePoint = UNICODE_REPLACEMENT_CHAR;
+                    }
+
+                    mUtf8Index = mUtf8ToFollow = 0;
+
+                    if (codePoint >= 0x80 && codePoint <= 0x9F) {
+                        // Sequence decoded to a C1 control character which we ignore. They are
+                        // not used nowadays and increases the risk of messing up the terminal state
+                        // on binary input. XTerm does not allow them in utf-8:
+                        // "It is not possible to use a C1 control obtained from decoding the
+                        // UTF-8 text" - http://invisible-island.net/xterm/ctlseqs/ctlseqs.html
+                    } else {
+                        switch (Character.getType(codePoint)) {
+                            case Character.UNASSIGNED:
+                            case Character.SURROGATE:
+                                codePoint = UNICODE_REPLACEMENT_CHAR;
+                        }
+                        processCodePoint(codePoint);
+                    }
+                }
+            } else {
+                // Not a UTF-8 continuation byte so replace the entire sequence up to now with the replacement char:
+                mUtf8Index = mUtf8ToFollow = 0;
+                emitCodePoint(UNICODE_REPLACEMENT_CHAR);
+                // The Unicode Standard Version 6.2 – Core Specification
+                // (http://www.unicode.org/versions/Unicode6.2.0/ch03.pdf):
+                // "If the converter encounters an ill-formed UTF-8 code unit sequence which starts with a valid first
+                // byte, but which does not continue with valid successor bytes (see Table 3-7), it must not consume the
+                // successor bytes as part of the ill-formed subsequence
+                // whenever those successor bytes themselves constitute part of a well-formed UTF-8 code unit
+                // subsequence."
+                processByte(byteToProcess);
+            }
+        } else {
+            if ((byteToProcess & 0b10000000) == 0) { // The leading bit is not set so it is a 7-bit ASCII character.
+                processCodePoint(byteToProcess);
+                return;
+            } else if ((byteToProcess & 0b11100000) == 0b11000000) { // 110xxxxx, a two-byte sequence.
+                mUtf8ToFollow = 1;
+            } else if ((byteToProcess & 0b11110000) == 0b11100000) { // 1110xxxx, a three-byte sequence.
+                mUtf8ToFollow = 2;
+            } else if ((byteToProcess & 0b11111000) == 0b11110000) { // 11110xxx, a four-byte sequence.
+                mUtf8ToFollow = 3;
+            } else {
+                // Not a valid UTF-8 sequence start, signal invalid data:
+                processCodePoint(UNICODE_REPLACEMENT_CHAR);
+                return;
+            }
+            mUtf8InputBuffer[mUtf8Index++] = byteToProcess;
+        }
+    }
+
+    public void processCodePoint(int b) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 252 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.